src/core/expression/qgsexpressioncontextutils.cpp

```diff
@@ -1224,6 +1268,32 @@ class CurrentFaceAreaExpressionFunction: public QgsScopedExpressionFunction
       if ( !context )
         return QVariant();
 
+      if ( context->hasVariable( QStringLiteral( "_mesh_face_index" ) ) && context->hasVariable( QStringLiteral( "_native_mesh" ) ) )
+      {
+        const int faceIndex = context->variable( QStringLiteral( "_mesh_face_index" ) ).toInt();
+        QgsMesh *nativeMesh = qvariant_cast<QgsMesh *>( context->variable( QStringLiteral( "_native_mesh" ) ) );
+        const QgsMeshFace &face = nativeMesh->face( faceIndex );
+        if ( !face.isEmpty() )
+        {
+          QgsDistanceArea *calc = parent->geomCalculator();
+          QgsGeometry geom = QgsMeshUtils::toGeometry( nativeMesh->face( faceIndex ), nativeMesh->vertices );
+          if ( calc )
+          {
+            double area = calc->measureArea( geom );
+            area = calc->convertAreaMeasurement( area, parent->areaUnits() );
+            return QVariant( area );
+          }
+          else
+          {
+            return QVariant( geom.area() );
+          }
+        }
+        else
+        {
+          return QVariant();
+        }
+      }
```

Comment:
In order to reduce code duplication in this file, I would suggest to have the code structured like this:
1. get native mesh - either from `_native_mesh` or from `_mesh_layer` (this could be even wrapped in a helper function, with a comment why there are two options) - or exit early with QVariant() on failure
2. do the actual work (extract vertex Z / calculate face area etc.)

Reply:
I completely removed the `_mesh_layer` handling and migrated the existing selection code to use the `_native_mesh` logic